File: app/servicos/textos.py
"""Normalizacoes textuais do dominio.

Regra transversal: nomes de posto, unidade, cargo e portaria sao gravados byte
a byte como aparecem no documento de origem. O sistema NUNCA normaliza caixa
nesses campos. As funcoes abaixo sao de *apresentacao* e de *comparacao*, nunca
de gravacao.
"""

from __future__ import annotations

import re
import unicodedata

# ---------------------------------------------------------------------
# UORG
# ---------------------------------------------------------------------
_RE_UORG = re.compile(r"^\s*(?P<codigo>\d+)\s*-\s*(?P<nome>.+?)\s*$")


def uorg_formatado(bruto: str | None) -> str:
    """'250 - FACULDADE DE MEDICINA DE DIAMANTINA'
        -> '250 - Faculdade De Medicina De Diamantina'

    Title case ingenuo, com o 'De' capitalizado. O assinado imprime assim -
    reproduzir exatamente, nao "melhorar" (CA-03).
    """
    if not bruto:
        return ""
    return str(bruto).title()


def codigo_uorg(bruto: str | None) -> str | None:
    """Extrai o codigo do prefixo '^(\\d+)\\s*-'."""
    if not bruto:
        return None
    m = _RE_UORG.match(str(bruto))
    return m.group("codigo") if m else None


def nome_uorg(bruto: str | None) -> str | None:
    if not bruto:
        return None
    m = _RE_UORG.match(str(bruto))
    return m.group("nome") if m else str(bruto).strip()


# ---------------------------------------------------------------------
# Aspas tipograficas
# ---------------------------------------------------------------------
def aspas_curvas(texto: str | None) -> str:
    """Converte aspas retas em curvas, alternando abre/fecha.

    Aplicado uma vez, na entrada do catalogo - nunca a cada render.
    """
    if not texto:
        return ""
    saida: list[str] = []
    abrindo = True
    for ch in str(texto):
        if ch == '"':
            saida.append("“" if abrindo else "”")
            abrindo = not abrindo
        else:
            saida.append(ch)
    return "".join(saida)


# ---------------------------------------------------------------------
# Comparacao / busca
# ---------------------------------------------------------------------
def sem_acento(texto: str | None) -> str:
    if not texto:
        return ""
    return "".join(
        c
        for c in unicodedata.normalize("NFD", str(texto))
        if unicodedata.category(c) != "Mn"
    )


def chave_busca(texto: str | None) -> str:
    """Minuscula, sem acento, espacos colapsados - para casar nomes."""
    return re.sub(r"\s+", " ", sem_acento(texto).lower()).strip()


def parece_cpf(bruto: str | None) -> bool:
    """Onze digitos seguidos so podem ser uma coisa nesta casa — e ela nao existe.

    O SIAPE tem sete (`ck_siape`), o protocolo de EPI tem prefixo, o NUP tem
    pontuacao e o numero de laudo tambem. Sobra o CPF, que o sistema NAO
    armazena em campo nenhum (RN-21; decisao 1 do desenho de EPI).

    Isto e deteccao, nao recusa: quem chama escreve a saida com as chaves da
    SUA tela. Mas a deteccao e uma so, aqui, porque duas copias divergem na
    primeira correcao — e a metade que ficasse para tras responderia "nada
    encontrado" a quem digitou CPF, deixando a pessoa concluir que o registro
    nao existe quando o que nao existe e a busca.
    """
    digitos = "".join(c for c in (bruto or "") if c.isdigit())
    return len(digitos) == 11


def normalizar_fluxo(texto: str) -> str:
    """Forma canonica de comparacao do CA-01.

    O PDF assinado quebra as linhas onde o texto envolve na pagina; o .docx tem
    paragrafos logicos. Comparar linha a linha compararia diagramacao, nao
    conteudo - por isso TODO espaco em branco (inclusive a quebra) e colapsado
    num unico espaco. A estrutura (ex.: a celula de posto ter exatamente duas
    linhas) e verificada em asseracoes proprias, nao aqui.
    """
    return re.sub(r"\s+", " ", normalizar_para_ouro(texto)).strip()


def normalizar_para_ouro(texto: str) -> str:
    """Normalizacao do CA-01: NFC + colapso de espacos + strip por linha.

    Nao compara bytes do .docx - compara texto normalizado.
    """
    texto = unicodedata.normalize("NFC", texto)
    texto = texto.replace(" ", " ").replace("\a", "\n").replace("\r\n", "\n")
    linhas = [re.sub(r"[ \t]+", " ", linha).strip() for linha in texto.split("\n")]
    return "\n".join(linha for linha in linhas if linha)


# ---------------------------------------------------------------------
# Nome de arquivo
# ---------------------------------------------------------------------
def slug_ascii(texto: str | None) -> str:
    """'Marco Antonio' -> 'Marco_Antonio'; 'Antônio' -> 'Antonio'.

    So o NOME DO ARQUIVO e ASCII; o conteudo do documento mantem acento.
    """
    base = sem_acento(texto or "")
    base = re.sub(r"[^A-Za-z0-9]+", "_", base).strip("_")
    return re.sub(r"_+", "_", base)


def nome_arquivo_parecer(numero: int, ano: int, sigla: str, servidor: str | None) -> str:
    """Parecer_Tecnico_08-2026_SEST_Talita - numero com padding de 2 no arquivo,
    sem padding no documento (n 8/2026)."""
    partes = ["Parecer_Tecnico", f"{numero:02d}-{ano}", slug_ascii(sigla.split("/")[0])]
    if servidor:
        partes.append(slug_ascii(servidor))
    return "_".join(p for p in partes if p)


# ---------------------------------------------------------------------
# RN-21 - dado de saude e proibido em campo de texto livre
#
# A lista abaixo e a regra. A unica flexao permitida esta em
# _DISPENSAS_POR_CONTEXTO, logo adiante, e vale por contexto declarado - nunca
# por frase.
# ---------------------------------------------------------------------
TERMOS_PROIBIDOS: tuple[str, ...] = (
    "cid-10",
    "cid 10",
    "diagnostico",
    "diagnóstico",
    "atestado medico",
    "atestado médico",
    "gestante",
    "gestacao",
    "gestação",
    "gravidez",
    "gravida",
    "grávida",
    "lactante",
    "lactacao",
    "lactação",
    "doenca",
    "doença",
    "enfermidade",
    "cpf",
)


# Contextos de aplicacao da RN-21.
#
# A regra nao e uniforme porque os campos nao sao. No parecer de adicional
# ocupacional nada de saude tem o que fazer: a insalubridade se prova pelo
# agente, pela concentracao e pelo ambiente - nunca pelo corpo do servidor. Ali
# a lista inteira vale, e continua valendo sem um milimetro de folga.
#
# Acidente em servico e outra coisa. A Lei 8.112/90, art. 212, e a decisao 4 do
# projeto definem tres especies, e uma delas se CHAMA "doenca relacionada ao
# trabalho". A palavra ali nao e diagnostico de ninguem: e o nome juridico da
# especie que se esta registrando, do mesmo naipe de "acidente de trajeto". Com
# a lista aplicada em bloco, o modulo de Acidentes nao conseguiria registrar a
# especie que a lei manda registrar.
#
# A dispensa e do CONTEXTO, nao de uma frase. Uma lista de frases liberadas
# ("doenca relacionada ao trabalho") quebra na primeira variacao de escrita e
# ainda deixa passar o texto clinico que venha grudado na frase liberada. O
# contexto e escolhido em cada ponto de chamada, aparece na revisao de codigo e
# se testa - e o default e sempre o mais restritivo, entao esquecer o parametro
# erra para o lado seguro.
CONTEXTO_PADRAO = "padrao"

# Vale so nos campos narrativos de `acidente_ocorrencia` e
# `acidente_investigacao`, onde a especie precisa ser nomeada. NAO vale em
# `acidente_decisao_pericial.observacao`: ali a tentacao e colar o laudo, que
# tem diagnostico, e o contexto continua sendo o CONTEXTO_PADRAO.
CONTEXTO_NEXO_OCUPACIONAL = "nexo_ocupacional"

# So a palavra que nomeia a especie legal. "enfermidade" segue barrada em toda
# parte: nao e nome de especie nenhuma, e quem precisa dela esta descrevendo
# quadro clinico. CID, diagnostico, atestado medico, gestacao e CPF seguem
# barrados em TODOS os contextos - nenhum deles nomeia nada na Lei 8.112.
#
# Risco residual assumido: com a palavra dispensada, "doenca de Chagas" passa no
# contexto do nexo. O filtro nunca foi o unico controle - o campo nasce
# restrito, com acesso auditado - e apertar mais so seria possivel voltando a
# casar frase, que e justamente o que nao funciona.
_DISPENSAS_POR_CONTEXTO: dict[str, frozenset[str]] = {
    CONTEXTO_PADRAO: frozenset(),
    CONTEXTO_NEXO_OCUPACIONAL: frozenset({"doenca", "doença"}),
}


class TextoProibido(ValueError):
    """Termo de saude/dado sensivel encontrado em campo de texto livre."""


def _dispensas_do_contexto(contexto: str) -> frozenset[str]:
    """Chaves de busca dispensadas no contexto. Contexto desconhecido e erro de
    programacao, e falha alto: um typo nao pode virar filtro silenciosamente
    diferente do que o autor quis."""
    if contexto not in _DISPENSAS_POR_CONTEXTO:
        conhecidos = ", ".join(sorted(_DISPENSAS_POR_CONTEXTO))
        raise ValueError(
            f"contexto de RN-21 desconhecido: {contexto!r}. Conhecidos: {conhecidos}."
        )
    return frozenset(chave_busca(t) for t in _DISPENSAS_POR_CONTEXTO[contexto])


def termos_proibidos_em(texto: str | None, contexto: str = CONTEXTO_PADRAO) -> list[str]:
    dispensadas = _dispensas_do_contexto(contexto)
    if not texto:
        return []
    alvo = chave_busca(texto)
    achados = []
    for termo in TERMOS_PROIBIDOS:
        chave = chave_busca(termo)
        if chave in dispensadas:
            continue
        if re.search(rf"\b{re.escape(chave)}\b", alvo):
            achados.append(termo)
    return sorted(set(achados))


def exigir_texto_limpo(
    texto: str | None, campo: str = "texto", contexto: str = CONTEXTO_PADRAO
) -> str | None:
    """Recusa o texto com dado de saude dentro. `campo` e o ROTULO DA TELA.

    Quem le esta frase e engenheiro de seguranca do trabalho, e a frase falava
    como banco de dados com ele: `'comentario'`, `'observacoes'`,
    `'justificativa_art9'` sao nomes de coluna, entre aspas, num sistema cujo
    resto do texto e portugues. O nome de coluna nao ajuda a achar o campo — a
    tela nao chama nenhum deles assim — e ainda expoe o esquema. Os pontos de
    chamada do modulo de EPI ja mandavam rotulo ("observacao da entrega",
    "parecer da analise"); os tres do Processos SEI eram os que faltavam.

    A recusa em si e deliberada e boa: comentario e observacao viram linha de
    auditoria, e auditoria nao guarda dado de saude sem base legal (RN-21; LGPD
    art. 11). O que faltava era a terceira parte da regra da casa — dizer a
    SAIDA. Ela existe, e e curta: descrever o agente e o ambiente resolve o
    trabalho tecnico inteiro sem nomear a condicao de ninguem.
    """
    achados = termos_proibidos_em(texto, contexto)
    if achados:
        raise TextoProibido(
            f"O campo {campo} contém termo proibido ({', '.join(achados)}). "
            "Estado de saúde, gestação, CID e diagnóstico não podem ser "
            "registrados (RN-21; LGPD art. 11). Descreva o agente nocivo e o "
            "ambiente de trabalho, não a pessoa — é isso que fundamenta o "
            "adicional, e o texto sem o termo vale igual."
        )
    return texto


# ---------------------------------------------------------------------
# Identificador opaco por sessao (RN-19)
# ---------------------------------------------------------------------
def identificador_opaco(servidor_id: int, semente: str) -> str:
    """SRV-7f3a - nunca mascara parcial da matricula."""
    import hashlib

    h = hashlib.sha256(f"{semente}:{servidor_id}".encode()).hexdigest()
    return f"SRV-{h[:4]}"

File: nuvem/src/servicos/textos.ts
/**
 * Normalizações textuais do domínio. Porte de `app/servicos/textos.py`.
 *
 * Regra transversal: nomes de posto, unidade, cargo e portaria são gravados
 * byte a byte como aparecem no documento de origem. O sistema NUNCA normaliza
 * caixa nesses campos. As funções abaixo são de *apresentação* e de
 * *comparação*, nunca de gravação.
 */
import { createHash } from "node:crypto";
import { RegraViolada } from "../nucleo/erros.js";

// ---------------------------------------------------------------------
// UORG
// ---------------------------------------------------------------------
const RE_UORG = /^\s*(\d+)\s*-\s*(.+?)\s*$/s;

function cased(c: string): boolean {
  return c.toLowerCase() !== c.toUpperCase();
}

/**
 * O `str.title()` do Python, fielmente: letra que vem depois de letra "com
 * caixa" vai para minúscula; qualquer outra vai para maiúscula. É ingênuo de
 * propósito ("De" capitalizado, "D'Agua"): o assinado imprime assim (CA-03).
 */
export function title_python(texto: string): string {
  let saida = "";
  let anteriorCased = false;
  for (const c of texto) {
    if (cased(c)) {
      saida += anteriorCased ? c.toLowerCase() : c.toUpperCase();
      anteriorCased = true;
    } else {
      saida += c;
      anteriorCased = false;
    }
  }
  return saida;
}

/**
 * '250 - FACULDADE DE MEDICINA DE DIAMANTINA'
 *   -> '250 - Faculdade De Medicina De Diamantina'
 *
 * Title case ingênuo, com o 'De' capitalizado. O assinado imprime assim —
 * reproduzir exatamente, não "melhorar" (CA-03).
 */
export function uorg_formatado(bruto: string | null | undefined): string {
  if (!bruto) return "";
  return title_python(String(bruto));
}

/** Extrai o código do prefixo '^(\d+)\s*-'. */
export function codigo_uorg(bruto: string | null | undefined): string | null {
  if (!bruto) return null;
  const m = RE_UORG.exec(String(bruto));
  return m ? m[1]! : null;
}

export function nome_uorg(bruto: string | null | undefined): string | null {
  if (!bruto) return null;
  const m = RE_UORG.exec(String(bruto));
  return m ? m[2]! : String(bruto).trim();
}

// ---------------------------------------------------------------------
// Aspas tipográficas
// ---------------------------------------------------------------------
/**
 * Converte aspas retas em curvas, alternando abre/fecha.
 * Aplicado uma vez, na entrada do catálogo — nunca a cada render.
 */
export function aspas_curvas(texto: string | null | undefined): string {
  if (!texto) return "";
  let saida = "";
  let abrindo = true;
  for (const ch of String(texto)) {
    if (ch === '"') {
      saida += abrindo ? "“" : "”";
      abrindo = !abrindo;
    } else {
      saida += ch;
    }
  }
  return saida;
}

// ---------------------------------------------------------------------
// Comparação / busca
// ---------------------------------------------------------------------
export function sem_acento(texto: string | null | undefined): string {
  if (!texto) return "";
  return String(texto).normalize("NFD").replace(/\p{Mn}/gu, "");
}

/** Minúscula, sem acento, espaços colapsados — para casar nomes. */
export function chave_busca(texto: string | null | undefined): string {
  return sem_acento(texto).toLowerCase().replace(/\s+/gu, " ").trim();
}

/**
 * Onze dígitos seguidos só podem ser uma coisa nesta casa — e ela não existe.
 *
 * O SIAPE tem sete, o protocolo de EPI tem prefixo, o NUP tem pontuação e o
 * número de laudo também. Sobra o CPF, que o sistema NÃO armazena (RN-21).
 * Isto é detecção, não recusa; mas a detecção é uma só, aqui.
 */
export function parece_cpf(bruto: string | null | undefined): boolean {
  const digitos = [...(bruto ?? "")].filter((c) => /\p{Nd}/u.test(c));
  return digitos.length === 11;
}

/**
 * Forma canônica de comparação do CA-01: TODO espaço em branco (inclusive a
 * quebra) é colapsado num único espaço — compara conteúdo, não diagramação.
 */
export function normalizar_fluxo(texto: string): string {
  return normalizar_para_ouro(texto).replace(/\s+/gu, " ").trim();
}

/** Normalização do CA-01: NFC + colapso de espaços + strip por linha. */
export function normalizar_para_ouro(texto: string): string {
  let t = texto.normalize("NFC");
  t = t.replaceAll(" ", " ").replaceAll("\x07", "\n").replaceAll("\r\n", "\n");
  const linhas = t.split("\n").map((linha) => strip_python(linha.replace(/[ \t]+/g, " ")));
  return linhas.filter((l) => l).join("\n");
}

/** `str.strip()` do Python (espaço em branco Unicode). */
function strip_python(s: string): string {
  return s.replace(/^[\s\x1c-\x1f\x85]+|[\s\x1c-\x1f\x85]+$/gu, "");
}

// ---------------------------------------------------------------------
// Nome de arquivo
// ---------------------------------------------------------------------
/**
 * 'Marco Antonio' -> 'Marco_Antonio'; 'Antônio' -> 'Antonio'.
 * Só o NOME DO ARQUIVO é ASCII; o conteúdo do documento mantém acento.
 */
export function slug_ascii(texto: string | null | undefined): string {
  let base = sem_acento(texto ?? "");
  base = base.replace(/[^A-Za-z0-9]+/g, "_").replace(/^_+|_+$/g, "");
  return base.replace(/_+/g, "_");
}

/**
 * Parecer_Tecnico_08-2026_SEST_Talita — número com padding de 2 no arquivo,
 * sem padding no documento (nº 8/2026).
 */
export function nome_arquivo_parecer(
  numero: number,
  ano: number,
  sigla: string,
  servidor: string | null | undefined,
): string {
  const partes = ["Parecer_Tecnico", `${String(numero).padStart(2, "0")}-${ano}`, slug_ascii(sigla.split("/")[0])];
  if (servidor) partes.push(slug_ascii(servidor));
  return partes.filter((p) => p).join("_");
}

// ---------------------------------------------------------------------
// RN-21 — dado de saúde é proibido em campo de texto livre
//
// A lista abaixo é a regra. A única flexão permitida está em
// DISPENSAS_POR_CONTEXTO, logo adiante, e vale por contexto declarado — nunca
// por frase.
// ---------------------------------------------------------------------
export const TERMOS_PROIBIDOS: readonly string[] = [
  "cid-10",
  "cid 10",
  "diagnostico",
  "diagnóstico",
  "atestado medico",
  "atestado médico",
  "gestante",
  "gestacao",
  "gestação",
  "gravidez",
  "gravida",
  "grávida",
  "lactante",
  "lactacao",
  "lactação",
  "doenca",
  "doença",
  "enfermidade",
  "cpf",
];

// Contextos de aplicação da RN-21 (ver o Python para o argumento inteiro): no
// parecer de adicional nada de saúde tem o que fazer; em Acidentes a Lei
// 8.112/90, art. 212, nomeia a espécie "doença relacionada ao trabalho". A
// dispensa é do CONTEXTO, não de uma frase, e o default é o mais restritivo.
export const CONTEXTO_PADRAO = "padrao";

// Vale só nos campos narrativos de `acidente_ocorrencia` e
// `acidente_investigacao`. NÃO vale em `acidente_decisao_pericial.observacao`.
export const CONTEXTO_NEXO_OCUPACIONAL = "nexo_ocupacional";

const DISPENSAS_POR_CONTEXTO: Record<string, readonly string[]> = {
  [CONTEXTO_PADRAO]: [],
  [CONTEXTO_NEXO_OCUPACIONAL]: ["doenca", "doença"],
};

/** Termo de saúde/dado sensível encontrado em campo de texto livre. */
export class TextoProibido extends RegraViolada {}

/**
 * Chaves de busca dispensadas no contexto. Contexto desconhecido é erro de
 * programação, e falha alto: um typo não pode virar filtro silenciosamente
 * diferente do que o autor quis.
 */
function dispensas_do_contexto(contexto: string): Set<string> {
  if (!Object.hasOwn(DISPENSAS_POR_CONTEXTO, contexto)) {
    const conhecidos = Object.keys(DISPENSAS_POR_CONTEXTO).sort().join(", ");
    throw new Error(`contexto de RN-21 desconhecido: '${contexto}'. Conhecidos: ${conhecidos}.`);
  }
  return new Set(DISPENSAS_POR_CONTEXTO[contexto]!.map((t) => chave_busca(t)));
}

function escapar_regex(s: string): string {
  // sem o `-`: fora de classe ele é literal, e `\-` é escape inválido no modo `u`
  return s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

export function termos_proibidos_em(texto: string | null | undefined, contexto: string = CONTEXTO_PADRAO): string[] {
  const dispensadas = dispensas_do_contexto(contexto);
  if (!texto) return [];
  const alvo = chave_busca(texto);
  const achados = new Set<string>();
  for (const termo of TERMOS_PROIBIDOS) {
    const chave = chave_busca(termo);
    if (dispensadas.has(chave)) continue;
    // `\b` do Python em str é Unicode; o do JS é só ASCII — daí os lookarounds
    const re = new RegExp(`(?<![\\p{L}\\p{N}_])${escapar_regex(chave)}(?![\\p{L}\\p{N}_])`, "u");
    if (re.test(alvo)) achados.add(termo);
  }
  return [...achados].sort();
}

/**
 * Recusa o texto com dado de saúde dentro. `campo` é o RÓTULO DA TELA, nunca
 * o nome da coluna. A recusa diz a SAÍDA: descrever o agente e o ambiente.
 */
export function exigir_texto_limpo<T extends string | null | undefined>(
  texto: T,
  campo = "texto",
  contexto: string = CONTEXTO_PADRAO,
): T {
  const achados = termos_proibidos_em(texto, contexto);
  if (achados.length) {
    throw new TextoProibido(
      `O campo ${campo} contém termo proibido (${achados.join(", ")}). ` +
        "Estado de saúde, gestação, CID e diagnóstico não podem ser " +
        "registrados (RN-21; LGPD art. 11). Descreva o agente nocivo e o " +
        "ambiente de trabalho, não a pessoa — é isso que fundamenta o " +
        "adicional, e o texto sem o termo vale igual.",
    );
  }
  return texto;
}

// ---------------------------------------------------------------------
// Identificador opaco por sessão (RN-19)
// ---------------------------------------------------------------------
/** SRV-7f3a — nunca máscara parcial da matrícula. */
export function identificador_opaco(servidor_id: number | null | undefined, semente: string): string {
  const h = createHash("sha256").update(`${semente}:${servidor_id ?? "None"}`, "utf8").digest("hex");
  return `SRV-${h.slice(0, 4)}`;
}
